### ISSUE TITLE
Incomplete arguments support

### ISSUE THREAD

**Issue by iliabylich** [sorbet/sorbet#1020]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0AA%20%3D%20T.type_alias(Integer)%0AB%20%3D%20T.type_alias(Integer)%0AC%20%3D%20T.type_alias(Integer)%0AD%20%3D%20T.type_alias(Integer)%0AE%20%3D%20T.type_alias(Integer)%0AF%20%3D%20T.type_alias(Integer)%0AG%20%3D%20T.type_alias(Integer)%0AH%20%3D%20T.type_alias(Integer)%0AI%20%3D%20T.type_alias(Integer)%0AJ%20%3D%20T.type_alias(Integer)%0AK%20%3D%20T.type_alias(Integer)%0A%0Asig%20do%0A%20%20params(%0A%20%20%20%20req%3A%20A%2C%0A%20%20%20%20opt%3A%20B%2C%0A%20%20%20%20rest%3A%20C%2C%0A%20%20%20%20post%3A%20D%2C%0A%20%20%20%20mlhs1%3A%20E%2C%0A%20%20%20%20mlhs2%3A%20F%2C%0A%20%20%20%20mlhs3%3A%20G%2C%0A%20%20%20%20mlhs4%3A%20H%2C%0A%20%20%20%20kwreq%3A%20I%2C%0A%20%20%20%20kwopt%3A%20J%2C%0A%20%20%20%20kwrest%3A%20T%3A%3AHash%5BSymbol%2C%20K%5D%2C%0A%20%20%20%20blk%3A%20T.proc.params(a%3A%20Integer%2C%20b%3A%20Integer).returns(Integer)%0A%20%20).void%0Aend%0Adef%20m(req%2C%20opt%20%3D%201%2C%20*rest%2C%20post%2C%20(mlhs1%2C%20*mlhs2%2C%20(mlhs3%2C%20mlhs4))%2C%20kwreq%3A%2C%20kwopt%3A%201%2C%20**kwrest%2C%20%26blk)%0Aend%0A%0A%0Am(1%2C%202%2C%203%2C4%2C%205%2C%20%5B6%2C%207%2C8%2C%20%5B9%2C10%5D%5D%2C%20kwreq%3A%2011%2C%20kwopt%3A%2012%2C%20whatever%3A%20100)%20%7B%20%7Ca%2Cb%7C%20a%20%2B%20b%20%7D%0A)

```ruby
# typed: true
extend T::Sig

A = T.type_alias(Integer)
B = T.type_alias(Integer)
C = T.type_alias(Integer)
D = T.type_alias(Integer)
E = T.type_alias(Integer)
F = T.type_alias(Integer)
G = T.type_alias(Integer)
H = T.type_alias(Integer)
I = T.type_alias(Integer)
J = T.type_alias(Integer)
K = T.type_alias(Integer)

sig do
  params(
    req: A,
    opt: B,
    rest: C,
    post: D,
    mlhs1: E,
    mlhs2: F,
    mlhs3: G,
    mlhs4: H,
    kwreq: I,
    kwopt: J,
    kwrest: T::Hash[Symbol, K],
    blk: T.proc.params(a: Integer, b: Integer).returns(Integer)
  ).void
end
def m(req, opt = 1, *rest, post, (mlhs1, *mlhs2, (mlhs3, mlhs4)), kwreq:, kwopt: 1, **kwrest, &blk)
end


m(1, 2, 3,4, 5, [6, 7,8, [9,10]], kwreq: 11, kwopt: 12, whatever: 100) { |a,b| a + b }

```

#### Observed output

Getting a bunch of errors:

```
editor.rb:32: Unsupported rest args in destructure https://srb.help/3009
    32 |def m(req, opt = 1, *rest, post, (mlhs1, *mlhs2, (mlhs3, mlhs4)), kwreq:, kwopt: 1, **kwrest, &blk)
                                                 ^^^^^^

[snip]

editor.rb:32: Malformed sig. Type not specified for argument <destructure> https://srb.help/5003
    32 |def m(req, opt = 1, *rest, post, (mlhs1, *mlhs2, (mlhs3, mlhs4)), kwreq:, kwopt: 1, **kwrest, &blk)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

editor.rb:36: [Integer(6), Integer(7), Integer(8), [Integer(9), Integer(10)]] does not match Integer for argument rest https://srb.help/7002
    36 |m(1, 2, 3,4, 5, [6, 7,8, [9,10]], kwreq: 11, kwopt: 12, whatever: 100) { |a,b| a + b }
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

#### Expected behavior

Expected this signature to be valid. I see multiple issues here:
1. Looks like mlhs arguments are not supported
2. Rest argument somehow "eats" all arguments after
3. kwrest argument also "eats" all kwreq and kwopt arguments.

- - -

I understand that this signature covers all args rules (there's also a "shadow" argument, but literally nobody needs it) and so I expected it to fail, but cases 2 and are pretty important.

**Comment by iliabylich** [comment]
Interesting, removing mlhs and kwrest fixes it. Does sorbet support kwrest argument?

**Comment by XrXr** [comment]
> EDIT: from what I see by using reveal_type sorbet wraps type of the kwarg into T::Array (similar to a regular restarg), but of course it's not an array. Is it a bug?

Could you post a link to sorbet.run that demonstrate this?


Side note, it looks like by declaring `kwrest: T::Hash[Symbol, K]` in the sig, [kwrest is `T::Hash[Symbol, T::Hash[Symbol, Integer]]` in the body](https://sorbet.run/#extend%20T%3A%3ASig%0A%0Asig%20%7Bparams(kwrest%3A%20T%3A%3AHash%5BSymbol%2C%20Integer%5D).void%7D%0Adef%20foo(**kwrest)%0A%20%20T.reveal_type(kwrest)%0Aend%0A).

**Comment by iliabylich** [comment]
> Could you post a link to sorbet.run that demonstrate this?

Weird, but I can't reproduce it anymore, so most probably I did something wrong.

> Side note, it looks like by declaring kwrest: T::Hash[Symbol, K] in the sig, kwrest is T::Hash[Symbol, T::Hash[Symbol, Integer]] in the body.

Ah, so it's similar to restarg. For `*rest` I should define type of the array element, for `**kwrest` - type of the value element (because keys are symbols, obviously)

Sorbet reports that it doesn't support mlhs arguments and all other cases work properly.

Closing, sorry for the noise.
